### DISCUSSION TITLE
Custom User Function in Template

### DISCUSSION THREAD

**Discussion by sez11a** [lervag/wiki.vim#354] [Q&A] [answered]
It's me again. :-) 

It's a new month, and in my planning system that means I plan out what I want to do that month in a separate file. To create that file in wiki.vim, I decided to auto-create a link in my `daily.md` template using a custom user function as described in the docs. This way, whenever I create a new daily file, the link to that month's plan is right at the top, and I can `:vsplit` and have the day on the left and the month on the right (similar to what I used to do in Logseq). 

I have this super-simple global function in lua: 

```lua
function WikiMonthLink()
  return print(os.date("[%B %Y](/journal/%Y/%m/plan.md)"))
end
```

Here's my template: 

```markdown
# {name}

{{WikiMonthLink}}

## Schedule



## Actions



## Roles 



## Notes
```
Where the placeholder is in the template, I just have a blank space. I am, admittedly, a Lua beginner, but this seemed so simple I can't see how it's not working. What am I missing?

**Comment by lervag** [comment, marked as answer]
> It's me again. :-)

Hi :)

> It's a new month, and in my planning system that means I plan out what I want to do that month in a separate file. To create that file in wiki.vim, I decided to auto-create a link in my `daily.md` template using a custom user function as described in the docs. This way, whenever I create a new daily file, the link to that month's plan is right at the top, and I can `:vsplit` and have the day on the left and the month on the right (similar to what I used to do in Logseq).

Sounds like a good idea!

> I have this super-simple global function in lua:
> 
> ```lua
> function WikiMonthLink()
>   return print(os.date("[%B %Y](/journal/%Y/%m/plan.md)"))
> end
> ```
> 
> Here's my template: …

The problem is that the file-based template only works with Vimscript functions. To be honest, I don't know quite how to fix that, but I think there is a decent workaround: instead of using a file template, you can use a function template. E.g.:

```lua
vim.g.wiki_templates = {
  {
    match_re = ".*",
    source_func = function(ctx)
      vim.api.nvim_put({
        "# " .. ctx.name,
        "",
        os.date("[%B %Y](/journal/%Y/%m/plan.md)"),
        "",
        "# Schedule",
        "...",
      }, "l", false, false)
    end
  },
}
```

The concept here is that the source function must add the desired lines by use of e.g. `nvim_put` or `nvim_buf_set_lines`; or with `vim.fn.append()`. The `context` argument has the template data, such as the `name` field.

The other alternative, of course, is to use a Vimscript function.

**Reply by sez11a** [reply]
Hey; that's awesome! Just one note: I have several templates (with more to be added), so I had to put this one last as a catch-all match. But this method did let me change the first line to 

```lua
"# " .. os.date("%A, %B %d, %Y"),
```
which is what I really wanted (rather than just the name of the file) anyway. 

>  To be honest, I don't know quite how to fix that

I get this; I imagine trying to support both Vim and Neovim, when everything in the Neovim world seems to be moving to Lua, is a challenge. I recently moved my [config](https://github.com/sez11a/vimstar) from Vimscript to Lua. I like Lua; it's a lot more straightforward, but re-learning how to do things is a pain. I don't have everything moved over, but it's mostly there. 

I have to say, `wiki.vim` is fantastic (if you combine it with `lists.vim`). As I said, I'm a Tech Writer for a software company, so I work in Markdown every day. When I tried Vimwiki, it took over the Markdown filetype, so I had to disable that part of it and use the wiki syntax, which I hated. Wiki.vim keeps things nice, stable, and easy to use. I'm looking forward to diving deeper into it.

**Reply by lervag** [reply]
> Hey; that's awesome!

Glad to hear it works!

> Just one note: I have several templates (with more to be added), so I had to put this one last as a catch-all match.

Yes; if you use several templates, you _will_ need to consider the order. The template objects are tried in the order they are provided, and only the first one that matches is used.

> But this method did let me change the first line to
> 
> ```lua
> "# " .. os.date("%A, %B %d, %Y"),
> ```

Neat. Notice, you could have done the same thing in Vimscript with `strftime()`.

> > To be honest, I don't know quite how to fix that
> 
> I get this; I imagine trying to support both Vim and Neovim, when everything in the Neovim world seems to be moving to Lua, is a challenge.

Yeah. I'll probably drop Vim support at some time, but it still feels fine to keep both. There are just a few cases where it's hard to make it fit nice right now.

> I recently moved my [config](https://github.com/sez11a/vimstar) from Vimscript to Lua. I like Lua; it's a lot more straightforward, but re-learning how to do things is a pain. I don't have everything moved over, but it's mostly there.

I also like Lua. It's a nice language, and with the type system from LSP I really do prefer it to Vimscript. Still, I find Vimscript is also quite nice, and much better than it's reputation.

> I have to say, `wiki.vim` is fantastic (if you combine it with `lists.vim`).

Thanks for the kind words! Don't hesitate to ask more questions or open issues.

**Reply by sez11a** [reply]
> Don't hesitate to ask more questions or open issues.

As you can see, I haven't. :-) 

Thanks for being so responsive. If you have any interest in my implementation, the templates are [here](https://github.com/sez11a/VimStar/tree/master/wiki/templates) and the config is [here](https://github.com/sez11a/VimStar/blob/master/lua/plugins/wiki.lua). I'm still building it out and (slowly) porting stuff over. I don't keep my personal info in GitHub; I have a Syncthing share where it goes to my personal computer, my work computer, my backup Raspberry Pi, my tablet, and my phone. Tablet has Termux and my Neovim setup; phone uses Markor. Everything's working well so far.

**Reply by lervag** [reply]
Cool! I looked at your config; good stuff. One thing that came to mind was that you might be interested in looking into snippet plugins. They are a general mechanism for templating that works for any file type. I use that for making templates for all kinds of things. I use [UltiSnips](https://github.com/SirVer/ultisnips), which is a classic, but there are also other and more modern variants.

I also use Syncthing, btw.; I find it works really well for sharing my personal files between trusted devices. I have Markor on my phone and use it once in a while, but not so often. Still, it is very useful to have the opportunity when I need it!

**Reply by lervag** [reply]
One thing, though: You might want to consider adjusting your match for the daily template to avoid having it match on any wiki file. For instance, you could use something like this:

```lua
          match_func = function(context)
            return context.path_wiki:find "%d%d%d%d/%d%d/%d%d"
          end,
```

**Reply by sez11a** [reply]
> You might want to consider adjusting your match for the daily template

That's a case of me not being sure the example you gave me above was a regex or a Lua pattern. :-)  I'll give the Lua pattern above a try; that's actually how I had it before the function, when I was using a template. 

I am using [Luasnip](https://github.com/L3MON4D3/LuaSnip) in my [completions.lua](https://github.com/sez11a/VimStar/blob/master/lua/plugins/completions.lua); I chose that because on Vim I used to use [Snipmate](https://github.com/garbas/vim-snipmate), and it supports all my old snippets. 

Since I embarked on this "go native with Neovim completions" project, I've tried to err on the side of Neovim-specific, Lua-based plugins, because that's what I see all the cool kids doing. Since Bram's death, it seems like that's now the future. :man_shrugging: 

Syncthing changed my life; I don't know what I'd do without it at this point. I don't like the cloud or streaming; I still buy CDs (mostly metal), rip them, and sync my music to my phone and use a local player. Pictures from the phone go to the computer; config files (like `gitconfig`, etc) sync to two computers and are symlinked to their proper locations. I love Syncthing.

**Reply by lervag** [reply]
> That's a case of me not being sure the example you gave me above was a regex or a Lua pattern. :-) I'll give the Lua pattern above a try; that's actually how I had it before the function, when I was using a template.

Yes, I can understand it is somewhat confusing. The value for `match_re` is a regular Vim regex, because it will be used in the context of Vimscript. If you use a Lua `match_func`, though, you must use a Lua pattern.

> I am using [Luasnip](https://github.com/L3MON4D3/LuaSnip) in my [completions.lua](https://github.com/sez11a/VimStar/blob/master/lua/plugins/completions.lua); I chose that because on Vim I used to use [Snipmate](https://github.com/garbas/vim-snipmate), and it supports all my old snippets.
> 
> Since I embarked on this "go native with Neovim completions" project, I've tried to err on the side of Neovim-specific, Lua-based plugins, because that's what I see all the cool kids doing. Since Bram's death, it seems like that's now the future. 🤷‍♂️

Yes, I've been thinking about testing Luasnip. But I've put it of, since things work well for me now. "If it ain't broke, ..." and all that. I have a mostly Lua based config myself, but I don't mind Vimscript and I have a very high investment in Vimscript through some of my plugins. Still, I agree that Neovim is the likely better future and I believe I will go more and more all in Lua with time.

> Syncthing changed my life; I don't know what I'd do without it at this point. I don't like the cloud or streaming; I still buy CDs (mostly metal), rip them, and sync my music to my phone and use a local player. Pictures from the phone go to the computer; config files (like gitconfig, etc) sync to two computers and are symlinked to their proper locations. I love Syncthing.

I also find Syncthing to be a key enabler technology. I was an early adoptor of Dropbox - it was a real game changer. But since I found Syncthing I've never been happier. It's just great!
